### ISSUE TITLE
Improve tests, ensure parsing and formatting regressions are caught

### ISSUE THREAD

**Issue by HarryR** [HarryR/logpipe#15]
Currently the test suite is sub-optimal, the [test.sh](test.sh) script provides some rudimentary checks, however I'm absolutely certain that I've subtly broken the `apacheclf` parser so that it rejects some lines which were previously successfully parsed.

Additionally the log files which are downloaded from random internet sites as part of the test suite provide no consistency, these files are useful because they provide a real log file sample and have highlighted a few issues (such as non-urlencoded spaces in the filename, and latin1 encoded characters), the `zzuf` tool is also useful in finding edge cases and has shown that the JSON serialiser used by `logpipe` can produce output that can't be parsed by PHP and Pythons native JSON modules.

The aims of a regression testing suite are:
- Given a specific pipeline configuration, with input and output samples, ensure that the pipeline exits with a specific known state.
- Ensure that known bugs are documented
- Ensure that known working pipelines continue to work
- Verify that changes to the pipeline and parsers don't cause regressions
- Verify that pipelines work consistently across platforms and build environments

I believe that when an appropriate regression testing framework is in place it will significantly improve confidence in the correctness of operation, and provide an easy way for the development cycle to incorporate examples of correct or incorrect behaviour from users.

In short, it lets us audit and verify the programs behaviour to achieve many goals and remain disciplined when developing features.

**Comment by HarryR** [comment]
The test handle for the pipeline should be very simple, it takes 4 input values, and returns success or failure.
- Input buffer
- Input pipeline
- Expected output buffer
- Expected output status code

Example:

``` c
logpipe_test(2, "parse.apacheclf print.apacheclf",
             "input line",
             "output line");
```

This will construct a new pipeline with the given steps, set the buffer to the input line, execute the pipeline, verify that 2 steps are executed, then check if the buffer matches the output line.

I will create a new unit test suite using minunit, during the process I will probably document many different bugs, and many working examples of different behaviours.

Then I will make sure that the unit tests are run on AppVeyor so functionality can be verified on Windows.

One bug I know will fail immediately is anon.randblank, the sequence it uses for pseudo-random numbers is deterministic, however it uses a static global to store the state, so the same pipeline constructed and executed twice in a row will give different results.
